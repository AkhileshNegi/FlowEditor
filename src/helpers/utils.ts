import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition } from '../flowTypes';

const request = require('sync-request');

const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase().trim(), SNAKED_CHARS, '_', 'all');
}

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export function getDisplayName(HOCName: string, Component: ComponentClass | SFC): string {
    const ComponentDisplayName = Component.displayName || Component.name || 'Component';
    return `${HOCName}(${ComponentDisplayName})`;
}

function getFlow(name: string): FlowDefinition {
    const definition = request('GET', 'base/test_flows/' + name + '.json').getBody();
    return JSON.parse(definition) as FlowDefinition;
}

export function getFavorites(): FlowDefinition {
    return getFlow('favorites');
}

export function getTest(): FlowDefinition {
    return getFlow('a4f64f1b-85bc-477e-b706-de313a022979');
}

export function dump(object: any) {
    console.log(JSON.stringify(object, null, 1));
}

export const DATA_SPEC_ATTRIBUTE_NAME: string = 'data-spec';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} matching DOM components
 */
export function getSpecWrapper(
    componentWrapper: ReactWrapper | ShallowWrapper,
    specName: string
): ShallowWrapper | ReactWrapper {
    return componentWrapper.find(`[${DATA_SPEC_ATTRIBUTE_NAME}="${specName}"]`);
}

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export function validUUID(uuid: string): boolean {
    const regex = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;
    return regex.test(uuid);
}
