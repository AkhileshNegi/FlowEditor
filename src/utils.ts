const XRegExp = require('xregexp');

const SNAKED_CHARS = XRegExp('[^\\p{Letter}\\d]+');

interface BoolMap {
    [key: string]: boolean;
}

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {BoolMap} A map of each string
 */
export function toBoolMap(array: string[]): BoolMap {
    return array.reduce((map: BoolMap, item: string) => {
        map[item] = true;
        return map;
    }, {});
}

export function addCommas(value: number): string {
    return value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');
}

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export function snakify(value: string): string {
    return XRegExp.replace(value.toLowerCase(), SNAKED_CHARS, '_', 'all');
}
