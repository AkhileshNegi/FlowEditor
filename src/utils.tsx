/** Turns a string array into a bool map */
export function toBoolMap(array: string[]): {[item: string]: boolean} {
    var map: {[item: string]: boolean} = {};
    for (let item of array) {
        map[item] = true;
    }
    return map;
}