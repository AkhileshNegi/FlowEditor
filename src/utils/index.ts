import { ComponentClass, SFC, ReactElement } from 'react';
import { ShallowWrapper, ReactWrapper } from 'enzyme';
import { FlowDefinition, Node, Languages, SwitchRouter, AnyAction } from '../flowTypes';
import { DragPoint } from '../component/Node';
import { PendingConnections, Components } from '../redux';
import { Language } from '../component/LanguageSelector';
import Localization, { LocalizedObject } from '../services/Localization';
import { NodeEditorProps } from '../component/NodeEditor';

const SNAKED_CHARS = /\s+(?=\S)/g;
const GRID_SIZE = 20;
export const V4_UUID = /^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i;

interface BoolMap {
    [key: string]: boolean;
}

interface Bounds {
    left: number;
    top: number;
    right: number;
    bottom: number;
}

/**
 * Adjusts the left and top offsets to a grid
 * @param left horizontal offset
 * @param top vertical offset
 */
export const snapToGrid = (left: number, top: number): { left: number; top: number } => {
    // Adjust our ghost into the grid
    let leftAdjust = left % GRID_SIZE;
    let topAdjust = top % GRID_SIZE;

    if (leftAdjust > GRID_SIZE / 3) {
        leftAdjust = GRID_SIZE - leftAdjust;
    } else {
        leftAdjust = leftAdjust * -1;
    }

    if (topAdjust > GRID_SIZE / 3) {
        topAdjust = GRID_SIZE - topAdjust;
    } else {
        topAdjust = topAdjust * -1;
    }

    return {
        left: Math.max(left + leftAdjust, GRID_SIZE * 2),
        top: Math.max(top + topAdjust, GRID_SIZE * 4)
    };
};

/**
 * Turns a string array into a bool map for constant lookup
 * @param {string[]} array - an array of strings, e.g. contact fields
 * @returns {object} A map of each string
 */
export const toBoolMap = (array: string[]): BoolMap =>
    array.reduce(
        (map: BoolMap, item: string) => ({
            ...map,
            [item]: true
        }),
        {}
    );

/**
 * Inserts commas into numbers where appropriate for better readability
 * @param {number} value - A number, e.g. 10000
 * @returns {string} A comma-separated string, e.g. 10,000
 */
export const addCommas = (value: number): string =>
    value.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ',');

/**
 * Replaces spaces with underscores
 * @param {string} value - A space-separated string to be snaked, e.g. a raw flow field name ('my flow field')
 * @returns {string} A snaked string, e.g. 'my_flow_field'
 */
export const snakify = (value: string): string =>
    value
        .toLowerCase()
        .trim()
        .replace(SNAKED_CHARS, '_');

/**
 * Returns a React component's name so we can attach it to a HOC's displayName property and view it in RDT
 * @param {ComponentClass | SFC} Component - A React component
 * @returns {string} The component's name
 */
export const getDisplayName = (WrappedComponent: ComponentClass | SFC): string =>
    WrappedComponent.displayName || WrappedComponent.name || 'Component';

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

/**
 * Returns true if a given UUID matches v4 format
 * @param {string} uuid - A version 4 UUID (no braces, uppercase OK)
 * @returns {boolean}
 */
export const validUUID = (uuid: string): boolean => V4_UUID.test(uuid);

/**
 * Returns a given string in title case, e.g. 'full name' becomes 'Full Name'
 * @param {string} str - string to be title-cased
 * @returns {string} Title-cased string
 */
export const titleCase = (str: string): string =>
    str.replace(/\b\w+/g, s => s.charAt(0).toUpperCase() + s.substr(1).toLowerCase());

export const getSelectClass = (errors: number): string => {
    if (errors === 0) {
        return '';
    }
    // We use a global selector here for react-select
    return 'select-invalid';
};

export const reorderList = (list: any[], startIndex: number, endIndex: number): any[] => {
    const [removed] = list.splice(startIndex, 1);

    list.splice(endIndex, 0, removed);

    return list;
};

/**
 * Compares basic objects (no methods and DOM nodes; property order important)
 * @param {object} objA - basic object
 * @param {object} objB - basic object
 * @returns {boolean}
 */
export const jsonEqual = (objA: {}, objB: {}): boolean =>
    JSON.stringify(objA) === JSON.stringify(objB);

/**
 * Checks whether any of a list of error strings contain one or more queries.
 * Used to determine whether a particular error has been encountered.
 */
export const hasErrorType = (errors: string[], exps: RegExp[]): boolean => {
    if (!errors.length) {
        return false;
    }
    for (const error of errors) {
        for (const exp of exps) {
            if (error.match(exp)) {
                return true;
            }
        }
    }
    return false;
};

export const pureSort = (list: any[], fn: (a: any, b: any) => number) => [...list].sort(fn);

export const getNodesBelow = ({ uuid: nodeUUID }: Node, nodes: Node[]) => {
    const idx = nodes.findIndex(({ uuid }: Node) => uuid === nodeUUID);
    return nodes.slice(idx, nodes.length);
};

export const getPendingConnection = (
    nodeUUID: string,
    pendingConnections: PendingConnections
): DragPoint => pendingConnections[nodeUUID];

export const getDetails = (uuid: string, components: Components) => components[uuid];

export const getNode = (uuid: string, components: Components, definition: FlowDefinition) => {
    const details = components[uuid];
    if (!details) {
        return null;
    }
    return definition.nodes[details.nodeIdx];
};

export const getExit = (uuid: string, components: Components, definition: FlowDefinition) => {
    const details = components[uuid];
    if (details) {
        const node = definition.nodes[details.nodeIdx];
        return node.exits[details.exitIdx];
    }
    return null;
};

export const getNodeUI = (uuid: string, definition: FlowDefinition) => definition._ui.nodes[uuid];

export const collides = (a: Bounds, b: Bounds) => {
    if (a.bottom < b.top || a.top > b.bottom || a.left > b.right || a.right < b.left) {
        return false;
    }
    // console.log("COLLISION!");
    return true;
};

export const getConnectionError = (source: string, targetUUID: string, components: Components) => {
    const { nodeUUID } = getDetails(source, components);
    if (nodeUUID === targetUUID) {
        return 'Connections cannot route back to the same places.';
    }
    return null;
};

/**
 * Computes translations prop for `Node` components in render()
 */
export const getTranslations = (definition: FlowDefinition, language: Language) => {
    if (definition.localization) {
        return definition.localization[language.iso];
    }
    return null;
};

export const getLocalizations = (
    node: Node,
    action: AnyAction,
    iso: string,
    languages: Languages,
    translations?: { [uuid: string]: any }
): LocalizedObject[] => {
    const localizations: LocalizedObject[] = [];

    // Account for localized cases
    if (node.router && node.router.type === 'switch') {
        const router = node.router as SwitchRouter;

        router.cases.forEach(kase =>
            localizations.push(Localization.translate(kase, iso, languages, translations))
        );

        // Account for localized exits
        if (node.exits) {
            node.exits.forEach(exit => {
                console.log('exit:', exit);
                localizations.push(Localization.translate(exit, iso, languages, translations));
            });
        }
    }

    if (action) {
        localizations.push(Localization.translate(action, iso, languages, translations));
    }

    return localizations;
};

export const determineConfigType = (
    nodeToEdit: Node,
    actions: AnyAction[],
    definition: FlowDefinition,
    components: Components
) => {
    if (actions.length) {
        return actions[actions.length - 1].type;
    } else {
        const nodeUI = definition._ui.nodes[nodeToEdit.uuid];
        if (nodeUI) {
            if (nodeUI.type) {
                return nodeUI.type;
            }
        }
    }

    // Account for ghost nodes
    if (nodeToEdit) {
        if (nodeToEdit.router) {
            return nodeToEdit.router.type;
        }

        if (nodeToEdit.actions) {
            return nodeToEdit.actions[0].type;
        }
    }

    const details = getDetails(nodeToEdit.uuid, components);
    if (details.type) {
        return details.type;
    }

    throw new Error(`Cannot initialize NodeEditor without a valid type: ${nodeToEdit.uuid}`);
};

export const getLocalizedObject = (localizations: LocalizedObject[]) => {
    if (localizations && localizations.length) {
        return localizations[0];
    }
};
