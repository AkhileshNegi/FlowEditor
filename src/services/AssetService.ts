// tslint:disable:max-classes-per-file
import { v4 as generateUUID } from 'uuid';
import { Endpoints, FlowEditorConfig, ContactProperties, Group } from '../flowTypes';
import axios, { AxiosResponse } from 'axios';
import { FlowComponents } from '../store/helpers';
import { dump } from '../utils';

export enum AssetType {
    Channel = 'channel',
    Flow = 'flow',
    Group = 'group',
    Property = 'property',
    Field = 'field'
}

export interface Asset {
    id: string;
    name: string;
    type: AssetType;

    isNew?: boolean;
    content?: any;
}

enum IdProperty {
    UUID = 'uuid',
    Key = 'key',
    ID = 'id'
}

enum NameProperty {
    Name = 'name'
}

enum SimAssetType {
    Flow = 'flow',
    Fields = 'field_set',
    Groups = 'group_set',
    Channels = 'channel_set'
}

interface SimAsset {
    type: SimAssetType;
    url: string;
    content: any;
}

export const getBaseURL = (): string => {
    const location = window.location;
    return (
        location.protocol + '//' + location.hostname + (location.port ? ':' + location.port : '')
    );
};

export class Assets {
    public endpoint: string;
    private localStorage: boolean;
    protected idProperty: IdProperty;
    protected assetType: AssetType;
    protected assets: { [id: string]: Asset } = {};

    constructor(endpoint: string, localStorage: boolean) {
        this.localStorage = localStorage;
        this.endpoint = endpoint;
        this.idProperty = IdProperty.UUID;
    }

    public get(id: string): Promise<Asset> {
        const existing = this.assets[id];
        if (existing) {
            return new Promise<Asset>(resolve => {
                resolve(existing);
            });
        }

        return new Promise<Asset>((resolve, reject) => {
            const url = `${this.endpoint}?uuid=${id}`;
            axios
                .get(url)
                .then((response: AxiosResponse) => {
                    const ob = response.data.results[0];
                    const asset = {
                        id: ob.uuid,
                        name: ob.name,
                        type: this.assetType,
                        content: ob.definition
                    };
                    if (this.localStorage) {
                        this.assets[id] = asset;
                    }
                    return resolve(asset);
                })
                .catch(error => reject(error));
        });
    }

    private searchLocalItems(term: string): Asset[] {
        // search our local items first
        const matches: Asset[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            if (this.matches(term, asset.name)) {
                matches.push(asset);
            }
        });
        return matches;
    }

    public search(term: string): Promise<Asset[]> {
        const matches = this.searchLocalItems(term);

        // then query against our endpoint to add to that list
        let url = this.endpoint;
        if (term) {
            url += '?query=' + encodeURIComponent(term);
        }

        return axios.get(url).then((response: AxiosResponse) => {
            for (const result of response.data.results) {
                if (this.matches(term, result.name)) {
                    matches.push({
                        name: result.name,
                        id: result[this.idProperty],
                        type: this.assetType
                    });
                }
            }
            return matches;
        });
    }

    public matches(query: string, check: string): boolean {
        if (!check) {
            return false;
        }

        if (query.length === 0) {
            return true;
        }

        return (
            check
                .toLocaleLowerCase()
                .trim()
                .indexOf(query.toLocaleLowerCase().trim()) > -1
        );
    }

    public add(result: Asset): void {
        if (this.localStorage) {
            this.assets[result.id] = result;
        }
    }

    public addAll(results: Asset[]): void {
        results.map((result: Asset) => {
            this.add(result);
        });
    }

    public update(uuid: string, content: any): Promise<Asset> {
        const result = new Promise<Asset>((resolve, reject) => {
            if (this.localStorage) {
                const asset = this.assets[uuid];
                asset.content = content;
                resolve(asset);
            } else {
                const url = `${this.endpoint}?uuid=${uuid}`;
                axios
                    .post(url, content)
                    .then((response: AxiosResponse) => {
                        const ob = response.data.results[0];
                        const asset = {
                            id: ob.uuid,
                            name: ob.name,
                            type: this.assetType,
                            content: ob.definition
                        };
                        if (this.localStorage) {
                            this.assets[uuid] = asset;
                        }
                        return resolve(asset);
                    })
                    .catch(error => reject(error));
            }
        });

        return result;
    }

    public getAssetSet(): any[] {
        const assets: any[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            if (asset.type !== AssetType.Property) {
                assets.push({
                    [this.idProperty]: asset.id,
                    name: asset.name,
                    ...asset.content
                });
            }
        });
        return assets;
    }

    public getAssetContents(type: SimAssetType): SimAsset[] {
        const assets: SimAsset[] = [];
        Object.keys(this.assets).map((key: string) => {
            const asset = this.assets[key];
            assets.push({
                type,
                url: getBaseURL() + this.endpoint + '/' + asset.id + '/',
                content: asset.content
            });
        });

        return assets;
    }
}

class GroupAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Group;
    }
}

class FieldAssets extends Assets {
    public static CONTACT_PROPERTIES: Asset[] = [
        {
            name: ContactProperties.Name,
            id: ContactProperties.Name.toLowerCase(),
            type: AssetType.Property
        }
        /*{ 
            name: ContactProperties.Language, 
            id: ContactProperties.Language.toLowerCase(), 
            type: AttributeType.property 
        }*/
    ];

    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.Key;
        this.assetType = AssetType.Field;

        FieldAssets.CONTACT_PROPERTIES.map((result: Asset) => {
            this.assets[result.id] = result;
        });
    }
}

class ChannelAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Channel;

        const uuid = '6934255e-446b-40d3-b2ca-c5f801bd2278';
        this.assets[uuid] = {
            id: uuid,
            name: 'Simulator',
            type: AssetType.Channel,
            content: {
                address: '+12065550000',
                schemes: ['tel'],
                roles: ['send', 'receive']
            }
        };
    }
}

class FlowAssets extends Assets {
    constructor(endpoint: string, localStorage: boolean) {
        super(endpoint, localStorage);
        this.idProperty = IdProperty.UUID;
        this.assetType = AssetType.Flow;
    }
}

export default class AssetService {
    private channels: Assets;
    private groups: Assets;
    private fields: Assets;
    private flows: FlowAssets;

    constructor(config: FlowEditorConfig) {
        this.groups = new GroupAssets(config.endpoints.groups, config.localStorage);
        this.fields = new FieldAssets(config.endpoints.fields, config.localStorage);
        this.flows = new FlowAssets(config.endpoints.flows, config.localStorage);

        // channels are always mocked for local
        this.channels = new ChannelAssets('/channels', true);
    }

    public addFlowComponents(flowComponents: FlowComponents): void {
        this.groups.addAll(flowComponents.groups);
        this.fields.addAll(flowComponents.fields);
    }

    public getFlowAssets(): Assets {
        return this.flows;
    }

    public getGroupAssets(): Assets {
        return this.groups;
    }

    public getFieldAssets(): Assets {
        return this.fields;
    }

    public getSimulationAssets(): any {
        let simAssets: SimAsset[] = [];
        const base = getBaseURL();
        // our group set asset
        simAssets.push({
            type: SimAssetType.Groups,
            url: base + this.groups.endpoint + '/',
            content: this.groups.getAssetSet()
        });

        // our fields
        simAssets.push({
            type: SimAssetType.Fields,
            url: base + this.fields.endpoint + '/',
            content: this.fields.getAssetSet()
        });

        // our channels
        simAssets.push({
            type: SimAssetType.Channels,
            url: base + this.channels.endpoint + '/',
            content: this.channels.getAssetSet()
        });

        // our flows
        simAssets = simAssets.concat(this.flows.getAssetContents(SimAssetType.Flow));

        const payload = {
            assets: simAssets,
            asset_server: {
                type_urls: {
                    flow: base + this.flows.endpoint + '/{uuid}/',
                    field_set: base + this.fields.endpoint + '/',
                    channel_set: base + this.channels.endpoint + '/',
                    group_set: base + this.groups.endpoint + '/'
                }
            }
        };

        return payload;
    }
}
