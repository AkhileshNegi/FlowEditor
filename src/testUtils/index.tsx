// TODO: Remove use of Function
// tslint:disable:ban-types
import { mount, ReactWrapper, shallow, ShallowWrapper } from 'enzyme';
import { object } from 'prop-types';
import * as React from 'react';
import * as config from '../../__test__/config';
import { ConfigProviderContext, endpointsPT, flowPT, languagesPT } from '../config';
import { FlowDefinition, FlowEditorConfig } from '../flowTypes';
import { AppState, createStore, initialState } from '../store';
import { getBaseLanguage } from '../utils';

export interface Resp {
    results: Array<{ [key: string]: any }>;
    total?: number;
}

export interface QueryString {
    [key: string]: string;
}

const initialContext: ConfigProviderContext = {
    endpoints: (config as FlowEditorConfig).endpoints,
    languages: (config as FlowEditorConfig).languages,
    flow: (config as FlowEditorConfig).flow
};

const initialTestState: AppState = {
    ...initialState,
    ...{
        flowContext: {
            ...initialState.flowContext,
            definition: require('../../__test__/flows/colors.json') as FlowDefinition
        }
    },
    ...{
        flowEditor: {
            ...initialState.flowEditor,
            editorUI: {
                ...initialState.flowEditor.editorUI,
                language: getBaseLanguage((config as FlowEditorConfig).languages)
            }
        }
    }
};

/**
 * Compose setup method for component tests
 */
export const createSetup = <P extends {}>(
    Component: React.ComponentClass | React.SFC,
    baseProps: P = {} as any,
    baseDuxState: AppState | Partial<AppState> = initialState,
    baseContext: ConfigProviderContext = initialContext,
    connectedChildren: boolean = false
) => (
    shallowRender: boolean = true,
    propOverrides: P | Partial<P> = {},
    duxStateOverrides = {} as AppState,
    contextOverrides = {} as ConfigProviderContext | Partial<ConfigProviderContext>,
    childContextTypeOverrides: { [key: string]: Function } = {}
) => {
    // Waiting on https://github.com/Microsoft/TypeScript/pull/1328
    const props = Object.assign({}, baseProps, propOverrides);
    const context = connectedChildren
        ? Object.assign({}, baseContext, {
              store: createStore(Object.assign({}, baseDuxState, duxStateOverrides))
          })
        : baseContext;
    // tslint:disable-next-line:ban-types
    const wrapper = (shallowRender ? (shallow as Function) : (mount as Function))(
        <Component {...props} />,
        {
            context,
            childContextTypes: Object.assign(
                {},
                {
                    store: object,
                    endpoints: endpointsPT,
                    languages: languagesPT,
                    flow: flowPT
                },
                childContextTypeOverrides
            )
        }
    );

    return {
        wrapper,
        props,
        context,
        instance: wrapper.instance()
    };
};
export const createSpy = (obj: Object | React.ComponentClass) => (instanceMethod: string) =>
    jest.spyOn((obj as React.ComponentClass).prototype || obj, instanceMethod as any);

/**
 * Wait for promises in queue to resolve
 */
export const flushPromises = () => new Promise(resolve => setImmediate(resolve));

/**
 * Restore spy mocks (distinct from mocks created w/ jest.fn()).
 * Use to declaratively restore multiple spy mocks,
 * otherwise just call .mockRestore() on lone spy.
 */
export const restoreSpies = (...spies: jest.SpyInstance[]) =>
    spies.forEach(spy => spy.mockRestore());

/**
 * NOTE: borrowed from EventBrite: https://github.com/eventbrite/javascript/blob/master/react/testing.md#finding-nodes
 * Finds all instances of components in the rendered `componentWrapper` that are DOM components
 * with the `data-spec` attribute matching `name`.
 * @param {ReactWrapper|ShallowWrapper} componentWrapper - Rendered componentWrapper (result of mount, shallow, or render)
 * @param  {string} snacName - Name of `data-spec` attribute value to find
 * @param {string|Function} (Optional) typeFilter - (Optional) Expected type of the wrappers (defaults to all HTML tags)
 * @returns {ReactWrapper|ReactWrapper[]|ShallowWrapper|ShallowWrapper[]} Matching DOM components
 */
export const getSpecWrapper = (
    componentWrapper: ReactWrapper<{}, {}> | ShallowWrapper<{}, {}>,
    specName: string,
    attributeName: string = 'data-spec'
): any => {
    return componentWrapper.find(`[${attributeName}="${specName}"]`);
};

const DEBUG = false;

export const timeStart = (name: string) => {
    if (DEBUG) {
        console.time(name);
    }
};

export const timeEnd = (name: string) => {
    if (DEBUG) {
        console.timeEnd(name);
    }
};
