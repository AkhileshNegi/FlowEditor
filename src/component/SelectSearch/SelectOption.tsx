// tslint:disable:ban-types
import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';

import { Asset } from '../../services/AssetService';
import { getIconForAssetType } from './helper';

export interface SelectOptionProps {
    className: string;
    isFocused: boolean;
    option: Asset;

    onFocus: Function;
    onSelect: Function;
}

export default class SelectOption extends React.PureComponent<SelectOptionProps, {}> {
    constructor(props: SelectOptionProps) {
        super(props);
        bindCallbacks(this, {
            include: [/^handle/]
        });
    }

    private handleMouseDown(event: any): void {
        event.preventDefault();
        event.stopPropagation();
        this.props.onSelect(this.props.option, event);
    }
    private handleMouseEnter(event: any): void {
        this.props.onFocus(this.props.option, event);
    }

    private handleMouseMove(event: any): void {
        if (this.props.isFocused) {
            return;
        }
        this.props.onFocus(this.props.option, event);
    }

    public render(): JSX.Element {
        return (
            <div
                className={this.props.className}
                onMouseDown={this.handleMouseDown}
                onMouseEnter={this.handleMouseEnter}
                onMouseMove={this.handleMouseMove}
            >
                <>
                    {getIconForAssetType(this.props.option.type)} {this.props.option.name}
                </>
            </div>
        );
    }
}
