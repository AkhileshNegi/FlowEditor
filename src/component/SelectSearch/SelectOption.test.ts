import { AssetType } from '../../services/AssetService';
import { composeComponentTestUtils, Resp } from '../../testUtils';
import { merge } from '../../utils';
import SelectOption, { SelectOptionProps } from './SelectOption';

const groupsResp = require('../../../__test__/assets/groups.json') as Resp;

const baseProps: SelectOptionProps = {
    className: 'myOption',
    isFocused: false,
    option: { name: 'Option', id: 'option-id', type: AssetType.Contact },
    onFocus: jest.fn(),
    onSelect: jest.fn()
};

const { setup, spyOn } = composeComponentTestUtils(SelectOption, baseProps);

describe(SelectOption.name, () => {
    describe('render', () => {
        it('should render user option', () => {
            const { wrapper } = setup(false, {
                $merge: {
                    option: {
                        name: 'My Contact',
                        id: 'contact-id',
                        type: AssetType.Contact
                    }
                }
            });

            expect(wrapper.html()).toMatchSnapshot('contact');
        });

        it('should render group option with icon', () => {
            const { wrapper } = setup(false, {
                $merge: {
                    option: {
                        name: 'My Group',
                        id: 'group-id',
                        type: AssetType.Group
                    }
                }
            });

            expect(wrapper.html()).toMatchSnapshot('group');
        });

        it('should render group option with icon', () => {
            const { wrapper } = setup(
                false,
                merge({
                    option: {
                        name: 'My URN',
                        id: 'tel:123456',
                        type: AssetType.URN
                    }
                })
            );

            expect(wrapper.html()).toMatchSnapshot('urn');
        });
    });

    describe('events', () => {
        it('handles mouse down', () => {
            const { wrapper, instance, props } = setup(false);
            wrapper.simulate('mouseDown', {});
            expect(props.onSelect).toHaveBeenCalled();
        });

        it('handles mouse move', () => {
            const { wrapper, instance, props } = setup(false);
            wrapper.simulate('mouseMove', {});
            expect(props.onFocus).toHaveBeenCalled();
        });

        it('handles mouse move focused', () => {
            const { wrapper, instance, props } = setup(false, { $merge: { isFocused: true } });
            wrapper.simulate('mouseMove', {});
            expect(props.onFocus).toHaveBeenCalled();
        });

        it('handles mouse enter', () => {
            const { wrapper, instance, props } = setup(false);
            wrapper.simulate('mouseEnter', {});
            expect(props.onFocus).toHaveBeenCalled();
        });
    });
});
