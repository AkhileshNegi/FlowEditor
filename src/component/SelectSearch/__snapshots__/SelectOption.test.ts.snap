// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectOption render should render group option with icon: group 1`] = `"<div class=\\"myOption\\"><div><span class=\\"fe-group\\"></span> My Group</div></div>"`;

exports[`SelectOption render should render group option with icon: urn 1`] = `"<div class=\\"myOption\\"><div> My URN</div></div>"`;

exports[`SelectOption render should render user option: contact 1`] = `"<div class=\\"myOption\\"><div> My Contact</div></div>"`;
